skaffold-plugins-core/src/main/java/com/google/cloud/tools/skaffold/command/Skaffold.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.skaffold.command;
+
+import com.google.cloud.tools.skaffold.downloader.SkaffoldDownloader;
+import com.google.cloud.tools.skaffold.filesystem.UserCacheHome;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.ByteStreams;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/** Runs {@code skaffold} commands. */
+public class Skaffold {
+
+  /** The location to store {@code skaffold} if auto-downloading it. */
+  private static final Path CACHED_SKAFFOLD_LOCATION =
+      UserCacheHome.getCacheHome().resolve("skaffold");
+
+  /** The location to store the digest for {@code skaffold} if auto-downloading it. */
+  private static final Path CACHED_SKAFFOLD_DIGEST_LOCATION =
+      CACHED_SKAFFOLD_LOCATION.resolveSibling("skaffold.sha256");
+
+  /**
+   * Initializes {@link Skaffold} with a custom path to the {@code skaffold} executable.
+   *
+   * @param executablePath the path to {@code skaffold}
+   * @return a new {@link Skaffold}
+   */
+  public static Skaffold atPath(Path executablePath) {
+    return new Skaffold(executablePath.toString());
+  }
+
+  /**
+   * Initializes {@link Skaffold} with a managed {@code skaffold} executable.
+   *
+   * @return a new {@link Skaffold}
+   * @throws IOException if an I/O exception occurred
+   */
+  public static Skaffold init() throws IOException {
+    ensureSkaffoldIsLatestVersion(CACHED_SKAFFOLD_LOCATION, CACHED_SKAFFOLD_DIGEST_LOCATION);
+    return new Skaffold(CACHED_SKAFFOLD_LOCATION.toString());
+  }
+
+  @VisibleForTesting
+  static void ensureSkaffoldIsLatestVersion(
```

Comment:
This method needs some logging, don't you think?

Reply:
Hmm, maybe it makes sense to break this out into a separate class that handles the auto-management.